src/mono/mono/mini/mini-llvm.c

```diff
@@ -8582,10 +8587,23 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 		}
 
 		case OP_SSE41_MUL: {
-			// NOTE: LLVM 7 and later use shifts here
-			// however, pmuldq is still available so I guess it's fine to keep using it
+#if LLVM_API_VERSION < 700
 			LLVMValueRef args [] = { lhs, rhs };
 			values [ins->dreg] = call_intrins (ctx, INTRINS_SSE_PMULDQ, args, dname);
+#else
+			const int shift_vals [] = { 32, 32 };
```

Comment:
Why duplicate the logic rather than just relying on the `upgradePMULDQ` path LLVM has internally, which seems like it would be better since it should get other bug fixes and improvements as time goes on, right?

Reply:
hm... looks like it makes sense

Reply:
Hmm, yeah; it looks like we could use llvm::UpgradeCallsToIntrinsic/llvm::UpgradeIntrinsicFunction/llvm::UpgradeIntrinsicCall to do this. But we already duplicate other vector-IR-generating logic from clang in the implementation of other intrinsics, so I think I'd prefer any refactoring that uses LLVM's auto-upgrade machinery to appear in a separate PR.

Reply:
I thought the point of `AutoUpgrade` is that you didn't have to make any changes. Rather existing IR should continue to "just work" and LLVM will internally change it when encountered to the IR supported by the newer version.

Reply:
We're generating LLVM IR in memory using LLVM's builder API and then either JIT compiling it with ORC or writing it to a bitcode file and using opt and llc to translate that bitcode file to an object file for AOT.

I'm looking through a local copy of LLVM; functions declared in AutoUpgrade.h are only ever used in BitcodeReader.cpp or LLParser.cpp. So auto-upgrading would "just work" for Mono + LLVM + AOT, because we're using opt and llc to generate native code, and these use IRReader.h functions, which themselves depend on functions declared in BitcodeReader.h and LLParser.h. But auto-upgrading doesn't look like it's performed automatically by ORC, and there's no parsing of serialized LLVM IR being done elsewhere as part of our LLVM JIT code, so we'd need to manually call these auto-upgrading functions somewhere else. Mono LLVM JIT and Mono LLVM AOT share almost all of the Mono IR to LLVM IR translation code, so adding auto-upgrading to Mono LLVM JIT in the most straightforward way I can think of will also change the IR we emit for AOT.

Also, we currently depend on intrinsic definitions generated by tablegen [[1]](https://github.com/dotnet/runtime/blob/5859dadf20a9dae44624e0a79e48f1b3f1f26a0a/src/mono/mono/mini/llvm-intrinsics.h) [[2]](https://github.com/dotnet/runtime/blob/5859dadf20a9dae44624e0a79e48f1b3f1f26a0a/src/mono/mono/mini/mini-llvm-cpp.cpp#L614). So we'd at the very least need an alternative way to declare LLVM intrinsic functions in each LLVM module we generate for those intrinsics that are recognized in llvm::UpgradeIntrinsicCall but have been removed from IntrinsicsX86.td.

Reply:
Thanks for the explanation, that makes sense to me!